src/Discord/Internal/Types/ApplicationCommands.hs

```diff
@@ -696,3 +797,74 @@ instance ToJSON ApplicationCommandPermissions where
         "type" .== applicationCommandPermissionsType,
         "permission" .== applicationCommandPermissionsPermission
       ]
+
+data Locale
+  = Da | De | EnGB | EnUS | EsES | Fr | Hr | It | Lt | Hu | Nl | No | Pl | PtBR
+  | Ro | Fi | SvSE | Vi | Tr | Cs | El | Bg | Ru | Uk | Hi | Th | ZhCN | Ja
+  | ZhTW | Ko
+  deriving (Show, Read, Eq, Ord)
+
+type TextTranslation = (Locale, T.Text)
+
+newtype LocalizedText = LocalizedText [TextTranslation] deriving (Show, Read, Eq, Ord)
+
+instance ToJSON LocalizedText where
+  toJSON (LocalizedText lt) = object . map (\(l,t) -> (localeString l, toJSON t)) $ lt
+    where
+      localeString l = case l of
+        Da -> "da"
+        De -> "de"
+        EnGB -> "en-GB"
+        EnUS -> "en-US"
+        EsES -> "es-ES"
+        Fr -> "fr"
+        Hr -> "hr"
+        It -> "it"
+        Lt -> "lt"
+        Hu -> "hu"
+        Nl -> "nl"
+        No -> "no"
+        Pl -> "pl"
+        PtBR -> "pt-BR"
+        Ro -> "ro"
+        Fi -> "fi"
+        SvSE -> "sv-SE"
+        Vi -> "vi"
+        Tr -> "tr"
+        Cs -> "cs"
+        El -> "el"
+        Bg -> "bg"
+        Ru -> "ru"
+        Uk -> "uk"
+        Hi -> "hi"
+        Th -> "th"
+        ZhCN -> "zh-CN"
+        Ja -> "ja"
+        ZhTW -> "zh-TW"
+        Ko -> "ko"
+  
+instance FromJSON LocalizedText where
+  parseJSON = withObject "LocalizedText"
+    (\o -> do
+        maybeTranslations <- sequenceSecond
+                             . map (\(l, t) -> (l, o .:? t))
+                             $ [ (Da, "da"), (De, "de"), (EnGB, "en-GB")
+                               , (EnUS, "en-US"), (EsES, "es-ES"), (Fr, "fr")
+                               , (Hr, "hr"), (It, "it"), (Lt, "lt"), (Hu, "hu")
+                               , (Nl, "nl"), (No, "no"), (Pl, "pl")
+                               , (PtBR, "pt-BR"), (Ro, "ro"), (Fi, "fi")
+                               , (SvSE, "sv-SE"), (Vi, "vi"), (Tr, "tr")
+                               , (Cs, "cs"), (El, "el"), (Bg, "bg"), (Ru, "ru")
+                               , (Uk, "uk"), (Hi, "hi"), (Th, "th")
+                               , (ZhCN, "zh-CN"), (Ja, "ja"), (ZhTW, "zh-TW")
+                               , (Ko, "ko")
+                               ]
+        let translations = map (\(l, Just t) -> (l, t)) $filter (\x -> (snd x) /= Nothing) maybeTranslations
```

Comment:
with `import Data.Maybe` and `{-# LANGUAGE TupleSections #-}, to reduce unsafety and improve readability

```
let translations = mapMaybe (\(l, t) -> fmap (l,) t) $ filter (isNothing . snd) maybeTranslations
```

Reply:
technically what you've written isn't unsafe it's just a bit hard to read

Reply:
I wanted to do it by somehow extracting the Pair list directly from the object, but there is no way to do it while remaining agnostic on the aeson version (spare a bunch of preproc directives

Reply:
you can use `AesonKey`, defined in Discord.Types.Prelude?

Reply:
Not really, in Aeson 1.x `Object` is `HashMap AesonKey Value` whereas in Aeson 2.x it is `KeyMap (Map AesonKey Value), and the level of indirection added prevents it from being easily doable

Reply:
Well I could probably throw together an `objectToList :: Object -> [(AesonKey, Value)]` with implementation being dependent on Aeson version